server/controllers/documents.js

```diff
@@ -45,18 +46,42 @@ export async function savePatch(req, res) {
   }
 
   try {
-    const count = await prisma.version.count({ where: { documentId } });
+    // count existing patches
+    const patchCount = await prisma.version.count({
+      where: { documentId, isSnapshot: false },
+    });
 
+    // always save the incoming patch
     const version = await prisma.version.create({
       data: {
         documentId,
         userId,
         diff: delta,
-        versionNumber: count + 1,
+        versionNumber: patchCount + 1,
         isSnapshot: false,
       },
     });
 
+    if ((patchCount + 1) % SNAPSHOT_INTERVAL === 0) {
+      await prisma.version.create({
+        data: {
+          documentId,
+          userId,
+          diff: fullContent,
+          versionNumber: patchCount + 2,
+          isSnapshot: true,
+        },
+      });
+
+      await prisma.document.update({
+        where: { id: documentId },
+        data: {
+          content: fullContent,
+          updatedAt: new Date(),
+        },
+      });
+    }
```

Comment:
The whole "snapshot" concept is really good.  Make sure to document this here in comments as well as in your writeups about the project/self review etc...

Reply:
Thought for future: What is the right snapshot interval? How would we determine that?